lib/Models/GeoJsonCatalogItem.ts

```diff
@@ -0,0 +1,674 @@
+                        });
+                    }
+                });
+        }
+    }
+
+    setTraitsFromData(geoJson: any): Promise<void> {
+        // If this GeoJSON data is an object literal with a single property, treat that
+        // property as the name of the data source, and the property's value as the
+        // actual GeoJSON.
+        var numProperties = 0;
+        var propertyName;
+        for (propertyName in geoJson) {
+            if (geoJson.hasOwnProperty(propertyName)) {
+                ++numProperties;
+                if (numProperties > 1) {
+                    break; // no need to count past 2 properties.
+                }
+            }
+        }
+
+        var name;
+        if (numProperties === 1 && isDefined(propertyName)) {
+            name = propertyName;
+            geoJson = geoJson[propertyName];
+            this.name = name;
+        }
+
+        // Reproject the features if they're not already EPSG:4326.
+        const proj4ServiceBaseUrl = this.catalogItem.terria.configParameters
+            .proj4ServiceBaseUrl;
+        return when(reprojectToGeographic(geoJson, proj4ServiceBaseUrl), () => {
+            this.geoJsonData = geoJson;
+        });
+    }
+}
+
+export default class GeoJsonCatalogItem
+    extends LoadGeoJsonMixin(
+        UrlMixin(CatalogMemberMixin(Model(GeoJsonCatalogItemTraits)))
+    )
+    implements Mappable {
+    @observable
+    show: boolean = true;
+
+    @observable
+    dataSource?: GeoJsonDataSource;
+
+    constructor(id: string, terria: Terria) {
+        super(id, terria);
+        this.strata.set(
+            LoadGeoJsonMixin.loadGeoJsonStratumName,
+            new LoadGeoJsonStratum(this)
+        );
+    }
+
+    @computed
+    get mapItems() {
+        if (this.dataSource === undefined) {
+            return [];
+        }
+
+        this.dataSource.show = this.show;
+        return [this.dataSource];
+    }
+
+    loadMetadata(): Promise<void> {
+        return Promise.resolve();
+    }
+
+    loadData(): Promise<void> {
+        const loadGeoJsonStratum = <LoadGeoJsonStratum>(
+            this.strata.get(LoadGeoJsonMixin.loadGeoJsonStratumName)
+        );
+        return loadGeoJsonStratum
+            .loadGeoJson()
+            .then(() => this.buildDataSource(this.geoJsonData))
```

Comment:
I think the creation of the `DataSource` can be deferred until `mapItems` is accessed. Doing it here will not be reactive when styles are changed after the GeoJSON is loaded.

Reply:
Again, think "my mapItems are a function of my GeoJSON data", rather than "when I load GeoJSON, I need to create a data source".

Reply:
`buildDataSource` is asynchronous, so this will need `mapItems` to return a `DataSource | Promise<DataSource>'. This doable since, `DataSourceCollection.add` can also accept a Promise. 

Does the whole change sound ok?

Reply:
It's possible to construct a `GeoJsonDataSource` and add it to the `DataSourceCollection` and _then_ call its `load` method. I think that would be better because it avoids dealing with the possibility that a `DataSource` is actually a `Promise<DataSource>` all over the place.